gateway-service/src/main/java/org/zowe/apiml/gateway/services/ServicesInfoService.java

```diff
@@ -0,0 +1,248 @@
+                    return authentication.isEmpty() ? null : authentication;
+                })
+                .filter(Objects::nonNull)
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    private Map<String, ServiceInfo.Instances> getInstances(List<InstanceInfo> appInstances) {
+        return appInstances.stream()
+                .collect(Collectors.toMap(
+                        InstanceInfo::getInstanceId,
+                        instanceInfo -> ServiceInfo.Instances.builder()
+                                .status(instanceInfo.getStatus())
+                                .hostname(instanceInfo.getHostName())
+                                .ipAddr(instanceInfo.getIPAddr())
+                                .protocol(getProtocol(instanceInfo))
+                                .port(getPort(instanceInfo))
+                                .homePageUrl(instanceInfo.getHomePageUrl())
+                                .healthCheckUrl(getHealthCheckUrl(instanceInfo))
+                                .statusPageUrl(instanceInfo.getStatusPageUrl())
+                                .customMetadata(getCustomMetadata(instanceInfo.getMetadata()))
+                                .build()
+                ));
+    }
+
+    private List<ServiceInfo.ApiInfoExtended> filterByIdAndMajorVersion(List<ServiceInfo.ApiInfoExtended> completeList) {
+        if (completeList.isEmpty()) return Collections.emptyList();
+
+        completeList.sort(Comparator
+                .comparing(ServiceInfo.ApiInfoExtended::getApiId, Comparator.nullsFirst(Comparator.naturalOrder()))
+                .thenComparing(ServiceInfo.ApiInfoExtended::getVersion, Comparator.comparing(this::getVersion))
+        );
+
+        List<ServiceInfo.ApiInfoExtended> result = new ArrayList<>();
+        ServiceInfo.ApiInfoExtended lastApiInfo = null;
+        for (ServiceInfo.ApiInfoExtended apiInfo : completeList) {
+            if (lastApiInfo == null ||
+                    !apiInfo.getApiId().equals(lastApiInfo.getApiId()) ||
+                    (apiInfo.getApiId().equals(lastApiInfo.getApiId()) &&
+                            getVersion(apiInfo.getVersion()).getMajorVersion() > getVersion(lastApiInfo.getVersion()).getMajorVersion())
+            ) {
+                result.add(apiInfo);
+                lastApiInfo = apiInfo;
+            }
+        }
+
+        return result;
+    }
+
+    private String getGatewayUrl(String url, String serviceId, ServiceType type, RoutedServices routes) {
+        if (url == null) return null;
+
+        try {
+            return transformService.transformURL(
+                    type,
+                    serviceId,
+                    url,
+                    routes);
+        } catch (URLTransformationException e) {
+            return url;
+        }
+    }
+
+    private Version getVersion(String version) {
```

Comment:
What about replace this Version class, with class like:

import org.apache.commons.lang3.StringUtils;

public class Version implements Comparable<Version> {

    private String[] numbers;

    public Version(String version) {
        this.numbers = version.split("\\.");
    }

    public int getLength() {
        return numbers.length;
    }

    public String getNumber(int i) {
        if (i < 0) {
            throw new IllegalArgumentException("Index of number has to be equal or greater than zero")
        }

        if (i >= numbers.length) {
            return "0";
        }

        return numbers[i];
    }

    private int compare(String a, String b) {
        if (StringUtils.isNumeric(a) && StringUtils.isNumeric(b)) {
            return Integer.valueOf(a).compareTo(Integer.valueOf(b));
        }
        return a.compareTo(b);
    }

    @Override
    public int compareTo(Version o) {
        int length = Math.max(getLength(), o.getLength());
        for (int i = 0; i < length; i++) {
            int c = compare(getNumber(i), o.getNumber(i));
            if (c != 0) {
                return c;
            }
        }

        return 0;
    }
}

It will be easier to use (see parsing). Support N numbers, not just 3. It supports also alphabets characters in a version.

Reply:
Feel free to do this change ;-)